### ISSUE TITLE
Timeout with GET /cluster/manager/validation in big environments

### ISSUE THREAD

**Issue by davidjiglesias** [wazuh/wazuh#6746]
Hello team,

As part of the investigation for https://github.com/wazuh/wazuh/issues/6744 we have discovered a possible timeout issue related with `GET /manager/configuration/validation` and `GET /cluster/configuration/validation`.

A fix to `wazuh-clusterd -t` has already been applied in https://github.com/wazuh/wazuh/issues/6744 but the cluster call was still taking too much time (6-9 seconds, very close to 10 seconds timeout) in a loaded cluster environment with 4 nodes and 40k agents.

We should investigate further and try to reduce this time. An alternative solution is to slightly increase default API (10s) and cluster (20s) timeouts, as those numbers are too low right now inside `cluster.json` file.

Regards,

David J. Iglesias

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

**Comment by Selutario** [comment]
## Issue update 

Hey team,

I have created an environment with 100,000 agents and 4 workers, which are synchronizing the information of 20,000 agents each. However, even though the conditions are similar or even worse than those mentioned in the issue, the `GET /manager/configuration/validation` and `GET /cluster/configuration/validation` endpoints still work correctly.

For example, I am getting times around 1.2-1.5 seconds when running `GET /cluster/configuration/validation`. 

I'll do a little more research in case I find anything suspicious that I may have missed. 

Regars,
Selu.

**Comment by Selutario** [comment]
## Issue update

The problem could appear in the past in large environments perhaps due to delays in wazuh-db, accumulated with the excessive time it took for wazuh-clusterd or wazuh-apid to test the configuration. 

However, as mentioned in the previous update, I can't reproduce it even with a higher number of agents (and the same cluster size). 

Regards,
Selu.

**Comment by Selutario** [comment]
## Issue update

Hey team,

I have done a more detailed analysis of which parts of the code are more time-consuming. 

To start with, the API is taking `0.743s` to respond to the `GET /manager/configuration/validation` endpoint. From that, `0.733s` corresponds to the execution of the SDK function that can be seen below: 
https://github.com/wazuh/wazuh/blob/f54e8e37ee03296c5bc5d323d85a1293caf10db0/framework/wazuh/manager.py#L193-L200

As could be expected, the slowest part of the SDK is the execution of the function `validate_ossec_conf()`. If we analyze the code of said function we can see that sending and receiving information from a socket is consuming `0.732s`, time measured in the following code block. This represents 98.5% of the endpoint response time. 
https://github.com/wazuh/wazuh/blob/f54e8e37ee03296c5bc5d323d85a1293caf10db0/framework/wazuh/core/manager.py#L142-L155

What the Wazuh core does when receiving the `check-manager-configuration` command that is sent in the previous socket is to run the following list of binaries with the `-t` option:
https://github.com/wazuh/wazuh/blob/f54e8e37ee03296c5bc5d323d85a1293caf10db0/src/os_execd/wcom.c#L321-L326

|Binary|Time spent|
|--|--|
|wazuh-authd -t|0.010s|
|wazuh-remoted -t|0.012s|
|wazuh-execd -t|0.006s|
|wazuh-analysisd -t|0.328s|
|wazuh-logcollector -t|0.007s|
|wazuh-integratord -t|0.009|
|wazuh-syscheckd -t|0.019s|
|wazuh-maild -t|0.008s|
|wazuh-modulesd -t|0.013s|
|wazuh-clusterd -t|0.240s|
|wazuh-agentlessd -t|0.011s|
|wazuh-integratord -t|0.007s|
|wazuh-dbd -t|0.013s|
|wazuh-csyslogd -t|0.005s|

The sum of all times is 0,68s, 91,52% of the total response time (`0.743s`). 

The ones that require the most time are `wazuh-clusterd` (`0.240s`) and `wazuh-analysisd` (`0.328s`). I have reviewed `wazuh-clusterd` although it does not seem feasible to further reduce the time it takes. Notice that around` 0.1s` is the time that the interpreter itself takes to execute the code. In addition, `0.15s` are then needed to import the modules (especially `cluster`) with the necessary tools to check the cluster configuration.
```
root@wazuh-master:/var/ossec# time bin/wazuh-clusterd -t
Time to import 'cluster': 0.120693s
Time to import 'cluster_utils': 0.000002s
Time to import 'pyDaemonModule': 0.000134s
Time to import 'common': 0.000002s
Time to import 'configuration': 0.000001s
```

Although the API is not included among the binaries whose configuration is checked with the X command, the time it takes for this task is high:
```
time bin/wazuh-apid -t
real	0m1.892s
user	0m1.811s
sys	0m0.078s
```

In this case, it is possible to reduce it. With this, we will not improve the endpoint that concerns us but it will take less time to start Wazuh. After the change, this is the time it takes to check the configuration: 
```
time bin/wazuh-apid -t

real	0m0.225s
user	0m0.193s
sys	0m0.033s
```

In conclusion, it seems impossible to significantly reduce the time it takes for said endpoint to respond. 

Regards,
Selu.

**Comment by mcarmona99** [comment]
## Issue update

In the [Wazuh 4.3.0 release candidate 5 testing protocol](https://github.com/wazuh/wazuh/issues/12521#issuecomment-1060309810), we have discovered that the error reported in this issue is still happening.

This time, the API timeout was given in the `PUT /cluster/{node_id}/configuration` endpoint. 

As it was said in this issue, the problem is related to the [connections with the `com` socket](https://github.com/wazuh/wazuh/issues/6746#issuecomment-861426617):

https://github.com/wazuh/wazuh/blob/11a2cea972e2976e663db1d1d130a10afdc04315/framework/wazuh/core/manager.py#L125-L153

As this socket is the same socket the `PUT /cluster/{node_id}/configuration` endpoint uses, the API timeout given there has the same origin as the API timeout that was first reported in this issue. 

As we can see, the `update_ossec_conf` function uses the `validate_ossec_conf`, which is the one used by the `GET /cluster/manager/validation` endpoint:

https://github.com/wazuh/wazuh/blob/11a2cea972e2976e663db1d1d130a10afdc04315/framework/wazuh/manager.py#L314-L316

This issue is reopening and in order to solve it, we should investigate deeper about connections with the com socket and these possible API timeouts.

<details>

<summary>test_cluster_endpoints result</summary>

```
============================= test session starts ==============================
platform linux -- Python 3.8.5, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
metadata: {'Python': '3.8.5', 'Platform': 'Linux-5.4.0-1029-aws-x86_64-with-glibc2.29', 'Packages': {'pytest': '5.4.3', 'py': '1.9.0', 'pluggy': '0.13.1'}, 'Plugins': {'tavern': '1.0.0', 'metadata': '1.10.0', 'html': '2.0.1'}}
rootdir: /tmp/Test_integration_endpoints_B2217_test_files/api/test/integration, inifile: pytest.ini
plugins: tavern-1.0.0, metadata-1.10.0, html-2.0.1
collecting ... collected 45 items

test_cluster_endpoints.tavern.yaml::GET /cluster/local/config PASSED     [  2%]
test_cluster_endpoints.tavern.yaml::GET /cluster/healthcheck PASSED      [  4%]
test_cluster_endpoints.tavern.yaml::GET /cluster/local/info PASSED       [  6%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes PASSED            [  8%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes[master-node] PASSED [ 11%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes[worker1] PASSED   [ 13%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes[worker2] PASSED   [ 15%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes select[ip] PASSED [ 17%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes select[name] PASSED [ 20%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes select[type] PASSED [ 22%]
test_cluster_endpoints.tavern.yaml::GET /cluster/nodes select[version] PASSED [ 24%]
test_cluster_endpoints.tavern.yaml::GET /cluster/status PASSED           [ 26%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/configuration PASSED [ 28%]
test_cluster_endpoints.tavern.yaml::GET /cluster/configuration/validation PASSED [ 31%]
test_cluster_endpoints.tavern.yaml::GET /cluster/api/config PASSED       [ 33%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/configuration/{component}/{configuration} PASSED [ 35%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/info[master-node] PASSED [ 37%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/info[worker1] PASSED [ 40%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/info[worker2] PASSED [ 42%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs[master-node] PASSED [ 44%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs[worker1] PASSED [ 46%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs[worker2] PASSED [ 48%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs/summary[worker1] PASSED [ 51%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/logs/summary[worker2] PASSED [ 53%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats[master-node] PASSED [ 55%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats[worker1] PASSED [ 57%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats[worker2] PASSED [ 60%]
test_cluster_endpoints.tavern.yaml::GET /cluster/wrong_node/stats PASSED [ 62%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/analysisd[master-node] PASSED [ 64%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/analysisd[worker1] PASSED [ 66%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/analysisd[worker2] PASSED [ 68%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/hourly[master-node] PASSED [ 71%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/hourly[worker1] PASSED [ 73%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/hourly[worker2] PASSED [ 75%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/remoted[master-node] PASSED [ 77%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/remoted[worker1] PASSED [ 80%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/remoted[worker2] PASSED [ 82%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/weekly[master-node] PASSED [ 84%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/weekly[worker1] PASSED [ 86%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/stats/weekly[worker2] PASSED [ 88%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/status[master-node] PASSED [ 91%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/status[worker1] PASSED [ 93%]
test_cluster_endpoints.tavern.yaml::GET /cluster/{node_id}/status[worker2] PASSED [ 95%]
test_cluster_endpoints.tavern.yaml::PUT /cluster/restart PASSED          [ 97%]
test_cluster_endpoints.tavern.yaml::PUT /cluster/{node_id}/configuration FAILED [100%]

=================================== FAILURES ===================================
_ /tmp/Test_integration_endpoints_B2217_test_files/api/test/integration/test_cluster_endpoints.tavern.yaml::PUT /cluster/{node_id}/configuration _
Format variables:
  protocol:s = 'https'
  host:s = 'localhost'
  port:d = '55000'
  valid_ossec_conf:s = '<ossec_config> <global> <jsonout_output>yes</jsonout_output> <alerts_log>yes</alerts_log> <logall>no</logall> <logall_json>no</logall_json> <email_notification>no</email_notification> <agents_disconnection_time>20s</agents_disconnection_time> <agents_disconnection_alert_time>100s</agents_disconnection_alert_time> </global>
<logging> <log_format>plain</log_format> </logging>
<alerts> <log_alert_level>300</log_alert_level> <email_alert_level>12</email_alert_level> </alerts>
<remote> <connection>secure</connection> <port>1514</port> <protocol>tcp</protocol> </remote>
<syscheck> <disabled>yes</disabled> </syscheck> </ossec_config>'
  test_login_token = 'eyJhbGciOiJFUzUxMiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ3YXp1aCIsImF1ZCI6IldhenVoIEFQSSBSRVNUIiwibmJmIjoxNjQ4MDM2OTU4LCJleHAiOjE2NDgwMzc4NTgsInN1YiI6InRlc3RpbmciLCJydW5fYXMiOmZhbHNlLCJyYmFjX3JvbGVzIjpbMV0sInJiYWNfbW9kZSI6IndoaXRlIn0.ABdsF4BC98Ar-3cnqMR10iHzplIMYfYeDrAKQTqNhxQsNmNaWR0B9Y5oSej5figygGvVqxPcopOqinGUyoofizSaAT7a7vbMmI5pqpIHqPHSnCeM0uXKKHw7hZjS070IPXsEJSE5WLEBFTesxiNDTmlv7AAUDAG_HTmKLRRetLy5Raw-'

Source test stage (line 1980):
  - name: Upload a valid configuration
    request:
      verify: False
      url: "{protocol:s}://{host:s}:{port:d}/cluster/worker1/configuration"
      method: PUT
      data: "{valid_ossec_conf:s}"
      headers:
        Authorization: "Bearer {test_login_token}"
        content-type: application/octet-stream
    response:
      status_code: 200
      json:
        data:
          affected_items:
            - 'worker1'
          failed_items: []
          total_affected_items: 1
          total_failed_items: 0
        error: 0



Formatted stage:
  name: Upload a valid configuration
  request:
    data: '<ossec_config> <global> <jsonout_output>yes</jsonout_output> <alerts_log>yes</alerts_log> <logall>no</logall> <logall_json>no</logall_json> <email_notification>no</email_notification> <agents_disconnection_time>20s</agents_disconnection_time> <agents_disconnection_alert_time>100s</agents_disconnection_alert_time> </global>
<logging> <log_format>plain</log_format> </logging>
<alerts> <log_alert_level>300</log_alert_level> <email_alert_level>12</email_alert_level> </alerts>
<remote> <connection>secure</connection> <port>1514</port> <protocol>tcp</protocol> </remote>
<syscheck> <disabled>yes</disabled> </syscheck> </ossec_config>'
    headers:
      Authorization: Bearer eyJhbGciOiJFUzUxMiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ3YXp1aCIsImF1ZCI6IldhenVoIEFQSSBSRVNUIiwibmJmIjoxNjQ4MDM2OTU4LCJleHAiOjE2NDgwMzc4NTgsInN1YiI6InRlc3RpbmciLCJydW5fYXMiOmZhbHNlLCJyYmFjX3JvbGVzIjpbMV0sInJiYWNfbW9kZSI6IndoaXRlIn0.ABdsF4BC98Ar-3cnqMR10iHzplIMYfYeDrAKQTqNhxQsNmNaWR0B9Y5oSej5figygGvVqxPcopOqinGUyoofizSaAT7a7vbMmI5pqpIHqPHSnCeM0uXKKHw7hZjS070IPXsEJSE5WLEBFTesxiNDTmlv7AAUDAG_HTmKLRRetLy5Raw-
      content-type: application/octet-stream
    method: PUT
    url: 'https://localhost:55000/cluster/worker1/configuration'
    verify: false
  response:
    json:
      data:
        affected_items:
        - worker1
        failed_items: []
        total_affected_items: 1
        total_failed_items: 0
      error: 0
    status_code: 200

Errors:
E   tavern.util.exceptions.TestFailError: Test 'Upload a valid configuration' failed:
    - Status code was 500, expected 200
    - Structure of returned data was different than expected  - Extra keys in response: {'detail', 'dapi_errors', 'title'} - Keys missing from response: {'data'} (expected = '{'data': {'affected_items': ['worker1'], 'failed_items': [], 'total_affected_items': 1, 'total_failed_items': 0}, 'error': 0}' (type = <class 'dict'>), actual = '{'title': 'Wazuh Internal Error', 'detail': 'Timeout executing API request', 'dapi_errors': {'worker1': {'error': 'Timeout executing API request', 'logfile': 'WAZUH_HOME/logs/cluster.log'}}, 'error': 3021}' (type = <class 'dict'>))
------------------------------ Captured log call -------------------------------
WARNING  tavern._plugins.rest.response:response.py:29 Unexpected status code '200'
ERROR    tavern.response.base:base.py:43 Status code was 500, expected 200
ERROR    tavern.response.base:base.py:41 Structure of returned data was different than expected  - Extra keys in response: {'detail', 'dapi_errors', 'title'} - Keys missing from response: {'data'} (expected = '{'data': {'affected_items': ['worker1'], 'failed_items': [], 'total_affected_items': 1, 'total_failed_items': 0}, 'error': 0}' (type = <class 'dict'>), actual = '{'title': 'Wazuh Internal Error', 'detail': 'Timeout executing API request', 'dapi_errors': {'worker1': {'error': 'Timeout executing API request', 'logfile': 'WAZUH_HOME/logs/cluster.log'}}, 'error': 3021}' (type = <class 'dict'>))
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 406, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert {'dapi_errors...ternal Error'} == {'data': {'af...}, 'error': 0}
  Differing items:
  {'error': 3021} != {'error': 0}
  Left contains 3 more items:
  {'dapi_errors': {'worker1': {'error': 'Timeout executing API request',
                               'logfile': 'WAZUH_HOME/logs/cluster.log'}},
   'detail': 'Timeout executing API request',
   'title': 'Wazuh Internal Error'}...
  
  ...Full output truncated (20 lines hidden), use '-vv' to show

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/tavern/response/base.py", line 88, in recurse_check_key_match
    check_keys_match_recursive(expected_block, block, [], strict)
  File "/usr/local/lib/python3.8/dist-packages/tavern/util/dict_util.py", line 447, in check_keys_match_recursive
    raise exceptions.KeyMismatchError(full_msg) from e
tavern.util.exceptions.KeyMismatchError: Structure of returned data was different than expected  - Extra keys in response: {'detail', 'dapi_errors', 'title'} - Keys missing from response: {'data'} (expected = '{'data': {'affected_items': ['worker1'], 'failed_items': [], 'total_affected_items': 1, 'total_failed_items': 0}, 'error': 0}' (type = <class 'dict'>), actual = '{'title': 'Wazuh Internal Error', 'detail': 'Timeout executing API request', 'dapi_errors': {'worker1': {'error': 'Timeout executing API request', 'logfile': 'WAZUH_HOME/logs/cluster.log'}}, 'error': 3021}' (type = <class 'dict'>))
=============================== warnings summary ===============================
/usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/hooks.py:83
  /usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/hooks.py:83: PytestDeprecationWarning: direct construction of YamlFile has been deprecated, please use YamlFile.from_parent
    return YamlFile(path, parent)

/usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:233: 23 tests with warnings
  /usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:233: PytestDeprecationWarning: direct construction of YamlItem has been deprecated, please use YamlItem.from_parent
    item = YamlItem(test_spec["test_name"], self, test_spec, self.fspath)

/usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:61
  /usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:61: PytestUnknownMarkWarning: Unknown pytest.mark.base_tests - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    pytest_marks.append(getattr(pytest.mark, m))

/usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:182: 33 tests with warnings
  /usr/local/lib/python3.8/dist-packages/tavern/testutils/pytesthook/file.py:182: PytestDeprecationWarning: direct construction of YamlItem has been deprecated, please use YamlItem.from_parent
    item_new = YamlItem(spec_new["test_name"], self, spec_new, self.fspath)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
- generated html file: file:///tmp/Test_integration_endpoints_B2217_test_files/api/test/integration/Test_integration_endpoints_B2217_test_cluster_endpoints.html -
=========================== short test summary info ============================
FAILED test_cluster_endpoints.tavern.yaml::PUT /cluster/{node_id}/configuration
============ 1 failed, 44 passed, 58 warnings in 639.17s (0:10:39) =============
```

</details>

[test_cluster_endpoints_logs.zip](https://github.com/wazuh/wazuh/files/8340473/Test_integration_endpoints_B2217_test_cluster_endpoints_logs.zip)

**Comment by davidjiglesias** [comment]
It looks like the problem comes from analysisd when it has to analyze trees with N depths. Not much we can do from API side.
